source/slang/slang-ir-witness-table-wrapper.cpp

```diff
@@ -0,0 +1,211 @@
+// slang-ir-witness-table-wrapper.cpp
+#include "slang-ir-witness-table-wrapper.h"
+
+#include "slang-ir-generics-lowering-context.h"
+#include "slang-ir.h"
+#include "slang-ir-clone.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+    struct GenericsLoweringContext;
+
+    struct GenerateWitnessTableWrapperContext
+    {
+        SharedGenericsLoweringContext* sharedContext;
+
+        IRStringLit* _getWitnessTableWrapperFuncName(IRFunc* func)
+        {
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedContext->sharedBuilderStorage;
+            builder->setInsertBefore(func);
+            if (auto linkageDecoration = func->findDecoration<IRLinkageDecoration>())
+            {
+                return builder->getStringValue((String(linkageDecoration->getMangledName()) + "_wtwrapper").getUnownedSlice());
+            }
+            if (auto namehintDecoration = func->findDecoration<IRNameHintDecoration>())
+            {
+                return builder->getStringValue((String(namehintDecoration->getName()) + "_wtwrapper").getUnownedSlice());
+            }
+            return nullptr;
+        }
+
+        IRFunc* emitWitnessTableWrapper(IRFunc* func, IRInst* interfaceRequirementVal)
+        {
+            auto funcTypeInInterface = cast<IRFuncType>(interfaceRequirementVal);
+
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedContext->sharedBuilderStorage;
+            builder->setInsertBefore(func);
+
+            auto wrapperFunc = builder->createFunc();
+            wrapperFunc->setFullType((IRType*)interfaceRequirementVal);
+            if (auto name = _getWitnessTableWrapperFuncName(func))
+                builder->addNameHintDecoration(wrapperFunc, name);
```

Comment:
Is it important for these wrappers to have distinguished names, or can we just leave them having the same name as the original, and then allow the emit pass to give them a unique suffix?

Reply:
Not that important, this is just to make it easier to identify a wrapper function.